### PR TITLE
0.10.0

### LINKED ISSUE #79: Higher precision decimals
So I'm going back and forth on the idea of [using a custom number type](https://github.com/maciejhirsz/json-rust/issues/60).

While it's nice to be able to parse higher precision than `f64`, it's also pretty useless unless there is a way to use that somehow. A solution for that could be methods like:
- `as_fixed_point_u64(u16) -> Option<u64>`
- `as_fixed_point_i64(u16) -> Option<i64>`

This way a number such as `12.3456` can be obtained as an integer `123456` when called with fixed point argument `4`. A way to construct such numbers would be needed as well (macro?)

Pros:
- Much faster serializing (canada.json will be a breeze).
- Parsing higher precision might be actually useful.

Cons:
- Obscuring the API, though this is already done with `Object` and `Short` types.
- Converting floats to the decimal representation is costly. I've already done an experimental alteration of the dtoa crate to produce decimal mantissa and exponent instead of writing to string with the grisu2 algorithm. Pefromance of that is quite okay.

Open questions:
- Is 64 bit mantissa for precision enough?
- Do we enable maths on the `Number` type and/or `JsonValue`?
- Can we use feature flags to convert numer types from other crates, in particular `num` crate?

### PR DESCRIPTION
**Pull request by maciejhirsz**
Closes #79

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -204,11 +204,14 @@ mod codegen;
...
 mod value;
 mod error;
+mod util;
 
 pub mod short;
 pub mod object;
+pub mod number;
 
 use short::Short;
+use number::Number;
```
Comment (Mark-Simulacrum) [inline src/lib.rs]: Maybe we should `pub use` here to re-export it? Although it seems that it'll rarely need to be used, so may not be worth it and little in the way of usability gains: `use json::number::Number` vs. `use json::Number`; however the second variant does bring similarity with `JsonValue`s.

File: src/number.rs [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+///
+/// More often than not you will deal with `JsonValue::Number` variant that
+/// wraps around this type, instead of using the methods here directly.
+#[derive(Copy, Clone, Debug)]
+pub struct Number {
+    // A byte describing the sign and NaN-ness of the number.
+    // category == 0 -> negative sign
+    // category == 1 -> positive sign
+    // category >  1 -> NaN
+    category: u8,
```
Comment (Mark-Simulacrum) [inline src/number.rs]: Might be worth mentioning the constants declared for this value at the top of the module.

File: src/parser.rs [hunk 3]
```diff
@@ -17,10 +17,13 @@
 // This makes for some ugly code, but it is faster. Hopefully in the future
 // with MIR support the compiler will get smarter about this.
 
-use std::{ ptr, mem, str, slice, char, f64 };
+use std::{ ptr, mem, str, slice, char };
 use object::Object;
+use number::Number;
 use { JsonValue, Error, Result };
 
+// This is not actual max precision, but a treshold at which number parsing
```
Comment (Mark-Simulacrum) [inline src/parser.rs]: `treshold` -> `threshold`.

File: src/lib.rs (hunk 1 above)
Comment (maciejhirsz) [inline src/lib.rs]: Yeah, analog to `Short` and `Object`, I'd prefer to "hide" the inner workings of the library. If you do - for whatever reason - want to use those as opposed to actual `JsonValue`, you do so explicitly.

I want to implement the arithmetic `ops` to the `Number` in future releases, re-exporting it at the top level then might make more sense.

Comment (Mark-Simulacrum) [inline src/lib.rs]: Sounds good. Just wanted to make sure that I mentioned it in case you wanted something different.

**Pull request merged**